service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,136 @@
+public class CertificateServiceImpl implements CertificateService {
+    private final MainRepository<Tag> tagRepository;
+    private final MainRepository<Certificate> repository;
+    private final CertificateValidator validator;
+
+    public CertificateServiceImpl(MainRepository<Certificate> repository,
+                                  MainRepository<Tag> tagRepository,
+                                  CertificateValidator validator) {
+        this.repository = repository;
+        this.tagRepository = tagRepository;
+        this.validator = validator;
+    }
+
+    @Override
+    public Certificate findById(long id) {
+        Specification<Certificate> specification
+                = Specification.of(new ModelByIdSpecification<>(id), new ModelNotRemovedSpecification<>());
+        Optional<Certificate> optionalCertificate = repository.queryFirst(specification);
+        return optionalCertificate.orElseThrow(EntityNotFoundException::new);
+    }
+
+    @Override
+    public Certificate update(Certificate certificate) {
+        throwExceptionIfNotValid(validator.validateForUpdate(certificate));
+
+        Optional<Certificate> optional = repository.queryFirst(new ModelByIdSpecification<>(certificate.getId()));
+        Certificate oldCertificate = optional.orElseThrow(EntityNotFoundException::new);
+
+        Certificate certificateToUpdate = Certificate.Builder.merge(oldCertificate, certificate);
+        List<Tag> ensuredTags = ensureTagsInRepo(certificateToUpdate.getTags());
+
+        certificateToUpdate = new Certificate.Builder(certificateToUpdate)
+                .setLastUpdateDate(LocalDate.now())
+                .setTags(ensuredTags)
+                .build();
+
+        return repository.update(certificateToUpdate);
+    }
+
+    @Override
+    public Certificate add(Certificate certificate) {
+        throwExceptionIfNotValid(validator.validateForCreate(certificate));
+
+        List<Tag> tags = CollectionUtils.isNotEmpty(certificate.getTags())
+                ? ensureTagsInRepo(certificate.getTags())
+                : Collections.emptyList();
+
+        certificate = new Certificate.Builder(certificate)
+                .setCreateDate(LocalDate.now())
+                .setLastUpdateDate(LocalDate.now())
+                .setTags(tags)
+                .build();
+
+        return repository.add(certificate);
+    }
+
+    private void throwExceptionIfNotValid(Map<String, String> validationFails) {
+        if (MapUtils.isNotEmpty(validationFails)) {
+            throw new ValidationException(validationFails);
+        }
+    }
+
+    private List<Tag> ensureTagsInRepo(List<Tag> tags) {
+        List<Tag> ensured = new ArrayList<>();
+
+        tags.forEach(tag -> {
+            Specification<Tag> specification = new TagByNameSpecification(tag.getName());
+            Optional<Tag> optional = tagRepository.queryFirst(specification);
+            tag = optional.isPresent() ? optional.get() : tagRepository.add(tag);
```

Comment:
Separated insert query for each tag. Check hibernate batch insert: [batch-insert](https://www.baeldung.com/jpa-hibernate-batch-insert-update)